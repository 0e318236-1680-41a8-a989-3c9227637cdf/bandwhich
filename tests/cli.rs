mod fakes;

use fakes::TerminalEvent::*;
use fakes::{
    get_interface, get_open_sockets, KeyboardEvents, NetworkFrames, TestBackend,
};

use ::insta::assert_snapshot;
use ::packet::builder::Builder;
use ::std::sync::{Arc, Mutex};
use ::termion::event::{Event, Key};

fn build_tcp_packet(
    source_ip: &str,
    destination_ip: &str,
    source_port: u16,
    destination_port: u16,
    payload: &'static [u8],
) -> Vec<u8> {
    ::packet::ether::Builder::default()
        .ip()
        .unwrap()
        .v4()
        .unwrap()
        .source(source_ip.parse().unwrap())
        .unwrap()
        .destination(destination_ip.parse().unwrap())
        .unwrap()
        .tcp()
        .unwrap()
        .source(source_port)
        .unwrap()
        .destination(destination_port)
        .unwrap()
        .payload(payload)
        .unwrap()
        .build()
        .unwrap()
}

struct LogWithMirror<T> {
    pub write: Arc<Mutex<T>>,
    pub mirror: Arc<Mutex<T>>,
}

impl<T> LogWithMirror<T> {
    pub fn new(log: T) -> Self {
        let write = Arc::new(Mutex::new(log));
        let mirror = write.clone();
        LogWithMirror { write, mirror }
    }
}

#[test]
fn basic_startup() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        None, // sleep
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![Clear, HideCursor, Draw, Flush, Clear, ShowCursor];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 1);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
}

#[test]
fn one_packet_of_traffic() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![Some(build_tcp_packet(
        "10.0.0.2",
        "1.1.1.1",
        443,
        12345,
        b"I am a fake tcp packet",
    ))]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn bi_directional_traffic() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "10.0.0.2",
            "1.1.1.1",
            443,
            12345,
            b"I am a fake tcp upload packet",
        )),
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I am a fake tcp download packet",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn multiple_packets_of_traffic_from_different_connections() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        Some(build_tcp_packet(
            "2.2.2.2",
            "10.0.0.2",
            54321,
            443,
            b"I come from 2.2.2.2",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn multiple_packets_of_traffic_from_single_connection() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I've come from 1.1.1.1 too!",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn one_process_with_multiple_connections() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        Some(build_tcp_packet(
            "3.3.3.3",
            "10.0.0.2",
            1337,
            443,
            b"Funny that, I'm from 3.3.3.3",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn multiple_processes_with_multiple_connections() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        Some(build_tcp_packet(
            "3.3.3.3",
            "10.0.0.2",
            1337,
            443,
            b"Awesome, I'm from 3.3.3.3",
        )),
        Some(build_tcp_packet(
            "2.2.2.2",
            "10.0.0.2",
            54321,
            443,
            b"You know, 2.2.2.2 is really nice!",
        )),
        Some(build_tcp_packet(
            "4.4.4.4",
            "10.0.0.2",
            1337,
            443,
            b"I'm partial to 4.4.4.4",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn multiple_connections_from_remote_ip() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12346,
            443,
            b"Me too, but on a different port",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 2);
    assert_snapshot!(&terminal_draw_events_mirror[0]);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
}

#[test]
fn sustained_traffic_from_one_process() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        None, // sleep
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"Same here, but one second later",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 3);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
    assert_snapshot!(&terminal_draw_events_mirror[2]);
}

#[test]
fn sustained_traffic_from_multiple_processes() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1",
        )),
        Some(build_tcp_packet(
            "3.3.3.3",
            "10.0.0.2",
            1337,
            443,
            b"I come from 3.3.3.3",
        )),
        None, // sleep
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"I have come from 1.1.1.1 one second later",
        )),
        Some(build_tcp_packet(
            "3.3.3.3",
            "10.0.0.2",
            1337,
            443,
            b"I come 3.3.3.3 one second later",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 3);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
    assert_snapshot!(&terminal_draw_events_mirror[2]);
}

#[test]
fn sustained_traffic_from_multiple_processes_bi_directional() {
    let keyboard_events = Box::new(KeyboardEvents::new(vec![
        None, // sleep
        None, // sleep
        None, // sleep
        Some(Event::Key(Key::Ctrl('c'))),
    ]));
    let network_frames = NetworkFrames::new(vec![
        Some(build_tcp_packet(
            "10.0.0.2",
            "3.3.3.3",
            443,
            1337,
            b"omw to 3.3.3.3",
        )),
        Some(build_tcp_packet(
            "3.3.3.3",
            "10.0.0.2",
            1337,
            443,
            b"I was just there!",
        )),
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"Is it nice there? I think 1.1.1.1 is dull",
        )),
        Some(build_tcp_packet(
            "10.0.0.2",
            "1.1.1.1",
            443,
            12345,
            b"Well, I heard 1.1.1.1 is all the rage",
        )),
        None, // sleep
        Some(build_tcp_packet(
            "10.0.0.2",
            "3.3.3.3",
            443,
            1337,
            b"Wait for me!",
        )),
        Some(build_tcp_packet(
            "3.3.3.3",
            "10.0.0.2",
            1337,
            443,
            b"They're waiting for you...",
        )),
        Some(build_tcp_packet(
            "1.1.1.1",
            "10.0.0.2",
            12345,
            443,
            b"1.1.1.1 forever!",
        )),
        Some(build_tcp_packet(
            "10.0.0.2",
            "1.1.1.1",
            443,
            12345,
            b"10.0.0.2 forever!",
        )),
    ]);

    let terminal_events = LogWithMirror::new(Vec::new());
    let terminal_draw_events = LogWithMirror::new(Vec::new());

    let backend = TestBackend::new(terminal_events.write, terminal_draw_events.write);
    let network_interface = get_interface();

    let os_input = what::OsInput {
        network_interface,
        network_frames,
        get_open_sockets,
        keyboard_events,
    };
    what::start(backend, os_input);

    let terminal_events_mirror = terminal_events.mirror.lock().unwrap();
    let terminal_draw_events_mirror = terminal_draw_events.mirror.lock().unwrap();

    let expected_terminal_events = vec![
        Clear, HideCursor, Draw, Flush, Draw, Flush, Draw, Flush, Clear, ShowCursor,
    ];
    assert_eq!(&terminal_events_mirror[..], &expected_terminal_events[..]);

    assert_eq!(terminal_draw_events_mirror.len(), 3);
    assert_snapshot!(&terminal_draw_events_mirror[1]);
    assert_snapshot!(&terminal_draw_events_mirror[2]);
}
