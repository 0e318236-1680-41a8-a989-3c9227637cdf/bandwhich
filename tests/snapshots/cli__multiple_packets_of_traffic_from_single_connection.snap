---
source: tests/cli.rs
expression: "&draw_log_copy.lock().unwrap()[0]"
---
[1;1H┌Utilization[1;14Hby[1;17Hconnection────────────────────────────────────────────────────────────────────┐┌Utilization[1;109Hby[1;112Hprocess[1;120Hname──────────────────────────────────────────────────────────────────┐[2;1H│[38;5;3mConnection[2;53HProcesses[2;74HTotal Bytes Up/Down[2;95H[39m││[38;5;3mProcess[2;128HConnection Count[2;159HTotal Bytes[2;190H[39m│[3;1H│[3;95H││[3;190H│[4;1H│[4;95H││[4;190H│[5;1H│[5;95H││[5;190H│[6;1H│[6;95H││[6;190H│[7;1H│[7;95H││[7;190H│[8;1H│[8;95H││[8;190H│[9;1H│[9;95H││[9;190H│[10;1H│[10;95H││[10;190H│[11;1H│[11;95H││[11;190H│[12;1H│[12;95H││[12;190H│[13;1H│[13;95H││[13;190H│[14;1H│[14;95H││[14;190H│[15;1H│[15;95H││[15;190H│[16;1H│[16;95H││[16;190H│[17;1H│[17;95H││[17;190H│[18;1H│[18;95H││[18;190H│[19;1H│[19;95H││[19;190H│[20;1H│[20;95H││[20;190H│[21;1H│[21;95H││[21;190H│[22;1H│[22;95H││[22;190H│[23;1H│[23;95H││[23;190H│[24;1H│[24;95H││[24;190H│[25;1H│[25;95H│└─────────────────────────────────────────────────────────────────────────────────────────────┘[26;1H│[26;95H│┌Utilization[26;109Hby[26;112Hremote[26;119Hip─────────────────────────────────────────────────────────────────────┐[27;1H│[27;95H││[38;5;3mRemote Address[27;148HConnection Count[27;169HTotal Bytes[27;190H[39m│[28;1H│[28;95H││[28;190H│[29;1H│[29;95H││[29;190H│[30;1H│[30;95H││[30;190H│[31;1H│[31;95H││[31;190H│[32;1H│[32;95H││[32;190H│[33;1H│[33;95H││[33;190H│[34;1H│[34;95H││[34;190H│[35;1H│[35;95H││[35;190H│[36;1H│[36;95H││[36;190H│[37;1H│[37;95H││[37;190H│[38;1H│[38;95H││[38;190H│[39;1H│[39;95H││[39;190H│[40;1H│[40;95H││[40;190H│[41;1H│[41;95H││[41;190H│[42;1H│[42;95H││[42;190H│[43;1H│[43;95H││[43;190H│[44;1H│[44;95H││[44;190H│[45;1H│[45;95H││[45;190H│[46;1H│[46;95H││[46;190H│[47;1H│[47;95H││[47;190H│[48;1H│[48;95H││[48;190H│[49;1H│[49;95H││[49;190H│[50;1H└─────────────────────────────────────────────────────────────────────────────────────────────┘└─────────────────────────────────────────────────────────────────────────────────────────────┘
